### PR TITLE
Collect Delta extended statistics when creating table

### PR DESCRIPTION
**Pull request by pajaks**
## Description
Collect delta lake statistics for CREATE TABLE AS.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Collect statistics for CREATE TABLE AS
```

### CODE REVIEW

**Comment by pajaks** [comment]
rebase on master to use CI fix https://github.com/trinodb/trino/pull/15879

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java [hunk 1]
```diff
@@ -68,6 +68,7 @@
     private Duration dynamicFilteringWaitTimeout = new Duration(0, SECONDS);
     private boolean tableStatisticsEnabled = true;
     private boolean extendedStatisticsEnabled = true;
+    private boolean collectExtendedStatisticsColumnStatisticsOnWrite = true;
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java:71]: maybe just `collectExtendedColumnStatisticsOnWrite` ?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 2]
```diff
@@ -2081,6 +2092,9 @@ public ConnectorAnalyzeMetadata getStatisticsCollectionMetadata(ConnectorSession
                     alreadyAnalyzedModifiedTimeMax.orElse(EPOCH)));
         }
 
+        Set<String> allColumnNames = extractColumnMetadata(metadata, typeManager).stream()
+                .map(ColumnMetadata::getName)
+                .collect(toImmutableSet());
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2097]: Save the result of `extractColumnMetadata` so that you don't have to call it again at the bottom of this method.
```suggestion
        List<ColumnMetadata> columnMetadata = extractColumnMetadata(metadata, typeManager);
        Set<String> allColumnNames = columnMetadata.stream()
                .map(ColumnMetadata::getName)
                .collect(toImmutableSet());
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 3]
```diff
@@ -2124,31 +2135,61 @@ public ConnectorAnalyzeMetadata getStatisticsCollectionMetadata(ConnectorSession
                 handle.getReadVersion(),
                 false);
 
+        TableStatisticsMetadata statisticsMetadata = getStatisticsCollectionMetadata(
+                statistics,
+                extractColumnMetadata(metadata, typeManager),
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2140]: Per other comment, don't have to call `extractColumnMetadata` again.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 4]
```diff
@@ -2124,31 +2135,61 @@ public ConnectorAnalyzeMetadata getStatisticsCollectionMetadata(ConnectorSession
...
+
+        Set<String> allColumnNames = tableMetadata.getColumns().stream()
+                .map(ColumnMetadata::getName)
+                .collect(toImmutableSet());
+
+        return getStatisticsCollectionMetadata(
+                Optional.empty(),
+                tableMetadata.getColumns(),
+                allColumnNames,
+                false);
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2162]: Why not `includeMaxFileModifiedTime` in this situation?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeConnectorSmokeTest.java [hunk 5]
```diff
@@ -1361,7 +1361,7 @@ private void testDeltaLakeTableLocationChanged(boolean fewerEntries, boolean fir
      * testing in {@link TestDeltaLakeAnalyze}.
      */
     @Test
-    public void testAnalyze()
+    public void testStatisticsGenerationDuringTableCreation()
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeConnectorSmokeTest.java:1364]: We should still test the old thing too

**Review by alexjo2144 — COMMENTED** [review]
Looks pretty good overall. Couple questions/nitpicks

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java [hunk 6]
```diff
@@ -147,22 +139,24 @@ private void testAnalyze(Optional<Integer> checkpointInterval)
     public void testAnalyzePartitioned()
     {
         String tableName = "test_analyze_" + randomNameSuffix();
-        assertUpdate("CREATE TABLE " + tableName
+        assertUpdate(
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java:142]: Please do create a compatibility test with spark to verify that after a CTAS `DESC EXTENDED` works as intended on Databricks

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java [hunk 7]
```diff
@@ -346,6 +347,19 @@ public DeltaLakeConfig setExtendedStatisticsEnabled(boolean extendedStatisticsEn
         return this;
     }
 
+    public boolean isCollectExtendedStatisticsColumnStatisticsOnWrite()
+    {
+        return collectExtendedStatisticsColumnStatisticsOnWrite;
+    }
+
+    @Config("delta.extended-statistics.collect-on-write")
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java:355]: Do consider documenting this new property in `delta-lake.rst` - either in this PR or a follow-up PR

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 4 above)
Comment (pajaks) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2162]: Statistic aggregation during table creation does not have information about file_modified_time yet.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java (hunk 7 above)
Comment (pajaks) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java:355]: I would wait with documentation until other write operations are implemented if that's ok.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 4 above)
Comment (alexjo2144) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2163]: Right, right. Then if the modified time isn't present we just use the current time when the collection is done. Makes sense.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java [hunk 8]
```diff
@@ -388,8 +382,7 @@ public void testAnalyzeSomeColumns()
     @Test
     public void testDropExtendedStats()
     {
-        try (TestTable table = new TestTable(
-                getQueryRunner()::execute,
+        try (TestTable table = new TestTable(getQueryRunner()::execute,
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java:385]: nit: There's no need to change this line. I would revert.

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 9]
```diff
@@ -2219,18 +2269,17 @@ public void finishStatisticsCollection(ConnectorSession session, ConnectorTableH
...
         }
 
-        if (analyzeHandle.getColumns().isPresent() && !mergedColumnStatistics.keySet().equals(analyzeHandle.getColumns().get())) {
-            // sanity validation
-            throw new IllegalStateException(
-                    format("Unexpected columns in in mergedColumnStatistics %s; expected %s",
-                            mergedColumnStatistics.keySet(),
-                            analyzeHandle.getColumns().get()));
-        }
+        analyzeHandle.flatMap(AnalyzeHandle::getColumns).ifPresent(analyzeColumns -> {
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2272]: nit: this kind of cosmetic changes can be done in a separate commit.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 10]
```diff
@@ -2402,7 +2451,8 @@ private static Optional<Instant> getMaxFileModificationTime(Collection<ComputedS
                     }
                     return Optional.of(Instant.ofEpochMilli(unpackMillisUtc(TimestampWithTimeZoneType.TIMESTAMP_TZ_MILLIS.getLong(entry.getValue(), 0))));
                 })
-                .collect(onlyElement());
+                .collect(toOptional())
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2454]: separate commit

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java (hunk 6 above)
Comment (findinpath) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java:142]: Nevermind. Trino Delta Lake (on the storage layer) & Databricks (on the metastore properties) have outputs in different places.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 4 above)
Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2163]: Can you please add a code comment explaining this consideration?
What do we need to have this information available?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 10 above)
Comment (pajaks) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2454]: This change make sense only with this commit as it allows collection to have 0 elements. It should throw exception before.

**Comment by pajaks** [comment]
CI https://github.com/trinodb/trino/issues/12535, https://github.com/trinodb/trino/issues/15809

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java [hunk 11]
```diff
@@ -67,20 +68,11 @@ public void testAnalyzeWithCheckpoints()
...
-                        "('comment', null, null, 0.0, null, null, null)," +
-                        "('name', null, null, 0.0, null, null, null)," +
-                        "(null, null, null, null, 25.0, null, null)");
-
-        runAnalyzeVerifySplitCount(tableName, 1);
+        assertUpdate(
+                "CREATE TABLE " + tableName + " " +
+                        (checkpointInterval.isPresent() ? format(" WITH (checkpoint_interval = %s)", checkpointInterval.get()) : "") +
+                        " AS SELECT * FROM tpch.sf1.nation",
+                25);
```
Comment (findepi) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java:75]: nit: unrelated fmt change

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java [hunk 12]
```diff
@@ -147,22 +139,24 @@ private void testAnalyze(Optional<Integer> checkpointInterval)
...
         String tableName = "test_analyze_" + randomNameSuffix();
-        assertUpdate("CREATE TABLE " + tableName
+        assertUpdate(
+                "CREATE TABLE " + tableName
                 + " WITH ("
                 + "   partitioned_by = ARRAY['regionkey']"
                 + ")"
-                + "AS SELECT * FROM tpch.sf1.nation", 25);
+                + "AS SELECT * FROM tpch.sf1.nation",
+                25);
```
Comment (findepi) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java:148]: nit: unrelated fmt change

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java [hunk 13]
```diff
@@ -276,7 +270,7 @@ public void testAnalyzeWithFilesModifiedAfter()
     {
         String tableName = "test_analyze_" + randomNameSuffix();
 
-        assertUpdate("CREATE TABLE " + tableName + " AS SELECT * FROM tpch.sf1.nation", 25);
+        assertUpdate(disableStatisticsCollectionOnWrite(getSession()), "CREATE TABLE " + tableName + " AS SELECT * FROM tpch.sf1.nation", 25);
```
Comment (findepi) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java:273]: nit: each arg on separate line

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java [hunk 14]
```diff
@@ -479,6 +469,70 @@ public void testStatsOnTpcDsData()
...
+        assertQuery(
+                "SHOW STATS FOR " + tableName,
+                "VALUES " +
+                        "('nationkey', null, null, 0.0, null, 0, 24)," +
+                        "('regionkey', null, 5.0, 0.0, null, null, null)," +
+                        "('comment', null, null, 0.0, null, null, null)," +
+                        "('name', null, null, 0.0, null, null, null)," +
+                        "(null, null, null, null, 25.0, null, null)");
+
+        runAnalyzeVerifySplitCount(tableName, 5);
```
Comment (findepi) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java:524]: i know it's preexisting but i don't think we need to assert split count in every test method here. It blurs the test's intent
(perhaps, we don't need it in any test, i don't know, but i am not requesting any change to existing tests)

this would be better:

```
assertUpdate("ANALYZE " + tableName);
```

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
@pajaks @findinpath @alexjo2144 thank you, this is awesome!

**Review by findepi — APPROVED** [review]

**Comment by pajaks** [comment]
CI:
suite-7-non-generic https://github.com/trinodb/trino/issues/14441
suite-iceberg https://github.com/trinodb/trino/issues/9078
